### PR TITLE
Budget App (Final Version)

### PR DESCRIPTION
**Pull request by katarighe**
# Ruby-on-Rails Capstone Project
In this capstone project, I developed a budget app named SmartPocket Budget keeps track of all your expenses associated with categories.

**Highlights**

✅ Integration & Unit tests passing
✅ Live Demo [deployed](https://moyen-budget.onrender.com/)
✅ Link to the [presentation video](https://www.loom.com/share/25cd395985e94de5aae700c872a7fa5d)

## Here are the following features done in this project.

**Splash screen**

To enhance the user onboarding process, I have incorporated a visually appealing splash screen that prominently displays the app's name. For added convenience, the splash screen provides direct links to the sign-up and log-in pages, streamlining the user's journey into the app.

**Signup and Login pages**

The app now offers a user registration feature, requiring users to provide their full name, email address, and password. Existing users can seamlessly access their accounts by logging in with their registered email and password. To ensure a secure authentication process, appropriate validations and security measures have been implemented. Unauthorized access to pages requiring authentication has been effectively restricted.

**Homepage and Categories page**

Upon successful login, users are directed to a dynamic home page showcasing various categories. Each category is presented with a clear and concise name, an accompanying icon, and the total amount of transactions associated with it. For enhanced user experience, clicking or tapping on a category item seamlessly navigates the user to the corresponding transactions page. Additionally, a prominent "Add a new category" button has been incorporated at the bottom of the page to facilitate category creation.

**Transactions page**

Users can now effortlessly view a comprehensive list of transactions for the selected category. The transactions are displayed in a clear and organized manner, sorted in descending order based on their recency, ensuring that users can easily track their latest expenses. The total amount for each category is prominently displayed at the top of the page, providing users with a quick overview of their financial standing. To further streamline transaction creation, a conveniently placed "Add a new transaction" button has been added at the bottom of the page. For added convenience, users can easily navigate back to the home page by simply clicking on the intuitive "Back" button (<).

**Add a New Category page**

A user-friendly form has been implemented to facilitate the creation of new categories. To successfully create a new category, users are required to provide a descriptive category name and an accompanying icon. Upon saving the newly created category, users are seamlessly redirected to the home page. Alternatively, users can conveniently navigate back to the home page by clicking on the intuitive "Back" button (<).

**Add a New Transaction page**

Users can effortlessly create new transactions by completing a straightforward form that includes the transaction name (mandatory), amount (mandatory), and at least one associated category (mandatory). Upon successfully saving the transaction, users are seamlessly directed to the corresponding transactions page for that specific category. To facilitate easy navigation, a convenient "Back" button (<) is provided, allowing users to swiftly return to the transactions page for the selected category.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,195 @@
...
+- Update or delete existing transactions.
+- Create new categories to classify your expenses by category.
+- View the calculated total amount of expenses per category.
+- Update or delete existing categories.
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+## 🚀 Live Demo <a name="live-demo"></a>
+
+- The Live Demo is available [here](https://budget-app-trxc.onrender.com)
```
Comment (DammyShittu) [inline README.md]: - [ ] I get a 502 error when I visit your live demo, kindly fix it.

  image

File: README.md [hunk 2]
```diff
@@ -1 +1,195 @@
...
+- Video of the app is available [here](https://www.loom.com/share/25cd395985e94de5aae700c872a7fa5d)
```
Comment (DammyShittu) [inline README.md]: - [ ] Nice presentation, but you failed to highlight a piece of code in your editor that you found interesting or an app feature you felt proud of making.

File: app/views/purchases/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+      <%= f.collection_check_boxes :group_ids, current_user.groups, :id, :name, :include_hidden => false %>
+      <span>No matching category?</span>
+      <%= link_to "add a new category", new_group_path, data: { turbo: false }, class: "static"  %>
+    </div>
+
+    <div id="actions">
+      <%= f.submit "add transaction", data: { turbo: false }, class: "btn"  %>
+    </div>
+  <% end %>
+</main>
```
Comment (DammyShittu) [inline app/views/purchases/new.html.erb]: - [x] When I am on the new transaction page there should be a back arrow button on the navbar, just like the one on the new category page that takes the user to the transactions page.

  image

File: spec/requests/purchases_spec.rb [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+RSpec.describe 'Purchases', type: :request do
+  let(:group) { Group.create(user: @user, name: 'Food', icon: 'missing_avatar.png') }
+  let(:purchase) { Purchase.create(name: 'Apples', amount: 5, author: @user) }
+  let(:group_purchase) { GroupPurchase.create(group: group, purchase: purchase) }
+  let(:valid_attributes) { { 'name' => 'Bananas', 'amount' => 5, 'author' => @user, 'group_ids' => [group.id] } }
+
+  before :each do
+    @user = User.create(name: 'Tom', email: 'tom@example.com', password: 'topsecret')
+    @user.confirm
```
Comment (DammyShittu) [inline spec/requests/purchases_spec.rb]: - [ ] Most of your tests are failing. Kindly fix them.

  image

**Review by DammyShittu — CHANGES_REQUESTED** [review]
Hi @katarighe,

## Highlights
✔️ Nice creative design to fit the design

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/purchases/new.html.erb (hunk 3 above)
Comment (katarighe) [reply app/views/purchases/new.html.erb]: ✅ Fixed

File: README.md (hunk 1 above)
Comment (katarighe) [reply README.md]: ✅ Done

Now: https://moyen-budget.onrender.com/

File: spec/requests/purchases_spec.rb (hunk 4 above)
Comment (julie-ify) [inline spec/requests/purchases_spec.rb]: - [ ] Most of the tests within the files in the `spec/requests` folder, like this one is failing due to this `@user.confirm` line.  When I removed this line in all the `spec/requests` folder, all these 33 specs in the folder pass.

  image

File: spec/features/groups_edit_spec.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
+require 'rails_helper'
+
+RSpec.describe 'when opening the group\'s edit page', type: :feature do
+  before(:each) do
+    @user = User.create(name: 'Tom', email: 'tom@example.com', password: 'topsecret')
+    @user.confirm
```
Comment (julie-ify) [inline spec/features/groups_edit_spec.rb]: - [x] You also need to remove all the `@user.confirm` lines in all the spec files contained in the `spec/features` folder. After that, most of these features' specs still pass, but you still need to take a look at the failing ones and fix them.

  image

File: spec/features/purchases_show_spec.rb [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    @group1 = Group.create(user: @user, name: 'Food', icon: 'https://i.pravatar.cc/300?img=13')
+    @group2 = Group.create(user: @user, name: 'Cosmetics', icon: 'https://i.pravatar.cc/300?img=1')
+    @purchase1 = Purchase.create(name: 'Apples', amount: 6, author: @user, groups: [@group1, @group2])
+    @purchase2 = Purchase.create(name: 'Bananas', amount: 4, author: @user, groups: [@group1])
+    visit(group_purchase_path(@group1, @purchase1))
+  end
+
+  context 'shows the right content' do
+    it 'heading' do
+      expect(page).to have_content('details')
```
Comment (julie-ify) [inline spec/features/purchases_show_spec.rb]: - [ ] As an example, this spec is failing, due to a case sensitivity issue. There is "Details" on the page but you are checking for "details" so it fails.

  image

File: spec/features/purchases_new_spec.rb [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    it 'renders new page again' do
+      click_button('add transaction')
+      expect(page).to have_content('new transaction')
+      expect(page).to have_button('add transaction', type: 'submit')
+    end
+
+    it 'shows an error message' do
+      click_button('add transaction')
+      expect(page).to have_content('You must choose at least one category!')
+    end
```
Comment (julie-ify) [inline spec/features/purchases_new_spec.rb]: - [x] After taking the approach I suggested you should now have only 8 failing specs, which you still need to figure out and fix, unlike the previous 98 failing specs.

  image

**Review by julie-ify — CHANGES_REQUESTED** [review]
Hi @katarighe,

Good job so far!

## Highlights!
- Smooth design and creativity ✔️ 
- Good code organization and working app ✔️
- Descriptive README ✔️ 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/purchases_new_spec.rb (hunk 7 above)
Comment (katarighe) [reply spec/features/purchases_new_spec.rb]: ✅ Solved

File: spec/features/purchases_show_spec.rb (hunk 6 above)
Comment (katarighe) [reply spec/features/purchases_show_spec.rb]: ✅ Solved

File: spec/features/groups_edit_spec.rb (hunk 5 above)
Comment (katarighe) [reply spec/features/groups_edit_spec.rb]: ✅ Solved

**Comment by katarighe** [comment]
All unit and integration tests are passing well again.

image

File: spec/features/pages_splash_spec.rb [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    end
+
+    it 'redirects to the splash page' do
+      expect(page).to have_current_path(unauthenticated_root_path)
+    end
+
+    it 'shows the notice that the confirmation email was sent' do
+      expect(page).to have_content('A message with a confirmation link has been sent to your email address.')
+    end
+  end
```
Comment (Qoosim) [inline spec/features/pages_splash_spec.rb]: - [x]  Great job working on making all the previously failed tests passed. However, there is this particular test on your way to the celebration. Take it out your way and let's go for the celebration. :balloon: 

failed-2
failed-test

**Review by Qoosim — CHANGES_REQUESTED** [review]
#  *Status: Required Changes* :recycle: 

Hi @katarighe, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


## Highlights
-  Implemented some required changes. :heavy_check_mark: 
-  No linter errors. :heavy_check_mark: 
-  Good README file. :heavy_check_mark: 
-  Correct Gitflow. :heavy_check_mark: 



 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by katarighe** [comment]
All the unit & integration tests are passing well again, after removing the non-existant example.
image

**Review by mckent05 — APPROVED** [review]
Hi @katarighe, 

# STATUS: APPROVED ✔️ ✔️ 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations and well done! 🎉 
approved


## Highlights
- All linter checks are passing ✔️ ✔️ 
- Your PR is professional 👏 👏
- Good job completing your rails capstone project 👍 👍 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**